### PR TITLE
Add OWLv2

### LINKED ISSUE #26315: Adding OwlV2 checkpoint to Owl-vit model
### Feature request

Google scenic team has released checkpoint of OwlV2 in the below link

https://github.com/google-research/scenic/tree/main/scenic/projects/owl_vit#pretrained-checkpoints

Could you please integrate this to HF’s transfomers?

### Motivation

OwlV2 has better performance while having simliar backbone as Owl-vit.

Please refer to performance comparison as following (https://arxiv.org/pdf/2306.09683.pdf)

### Your contribution

I don’t think I can be a main contributer especially I don’t know model converting. However, I know A to Z of transformers’ Owl-vit code so I can do some job (assign me part of the work!)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the OWLv2 checkpoints, and investigates why results weren't the same as the original Colab for v1 as pointed out in #21206. After investigation, it turns out that 1) the model returns the exact same logits as the original implementation on the same input data, but 2) the image preprocessing wasn't done in exactly the same way as the original Scenic repo, which involves padding, causing a difference. This PR fixes that by introducing a new `Owlv2ImageProcessor` which people can use to get equivalent results. 

Fixes #26315 #21206

Question:
- the new OWLv2 models have an extra objectness head, which returns `objectness_logits` => we can either add an entirely new `Owlv2ForObjectDetection` which copies 99% of `OwlViTForObjectDetection`, or do it like it's currently implemented (by adding a `config.add_objectness_head` attribute). Happy to read your opinions!

To do:

- [ ] update layer norm eps of LayerNorms, which is 1e-5 in PyTorch but 1e-6 in Flax by default

### CODE REVIEW

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 1]
```diff
@@ -1699,6 +1720,11 @@ def forward(
         # Predict object classes [batch_size, num_patches, num_queries+1]
         (pred_logits, class_embeds) = self.class_predictor(image_feats, query_embeds, query_mask)
 
+        # Predict objectness
+        objectness_logits = None
+        if self.objectness_head is not None:
+            objectness_logits = self.objectness_predictor(image_feats)
+
```
Comment (ydshieh) [inline src/transformers/models/owlvit/modeling_owlvit.py:1727]: maybe add a comment mentioning:

`objectness_logits` is used to sample top tokens during training that is not currently supported. 

https://github.com/google-research/scenic/blob/a41317783b2f33f85b94f7af0860a93c4265db58/scenic/projects/owl_vit/models.py#L322-L333

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 2]
```diff
@@ -1723,4 +1750,5 @@ def forward(
             class_embeds=class_embeds,
             text_model_output=text_outputs,
             vision_model_output=vision_outputs,
+            objectness_logits=objectness_logits,
```
Comment (ydshieh) [inline src/transformers/models/owlvit/modeling_owlvit.py:1753]: Would be great to add a new example in the docstring of `OwlViTForObjectDetection`.

**Comment by ydshieh** [comment]
I would go for simply adding `config.add_objectness_head` in the existing `OwlViT` model file(s).

**Comment by ArthurZucker** [comment]
Sorry for the delay I am reviewing now 😉

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 3]
```diff
@@ -1711,6 +1737,7 @@ def forward(
                 class_embeds,
                 text_outputs.to_tuple(),
                 vision_outputs.to_tuple(),
+                objectness_logits,
```
Comment (ArthurZucker) [inline src/transformers/models/owlvit/modeling_owlvit.py:1740]: this is breaking 😅 
One way to prevent this from breaking is to have a `return_objectness_logits` set to false by default

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 4]
```diff
@@ -1325,7 +1328,8 @@ def __init__(self, config: OwlViTConfig):
 
         self.owlvit = OwlViTModel(config)
         self.class_head = OwlViTClassPredictionHead(config)
-        self.box_head = OwlViTBoxPredictionHead(config)
+        self.box_head = OwlViTBoxPredictionHead(config, out_dim=4)
+        self.objectness_head = OwlViTBoxPredictionHead(config, out_dim=1) if config.add_objectness_head else None
```
Comment (ArthurZucker) [inline src/transformers/models/owlvit/modeling_owlvit.py:1332]: Mmmmm this will change the forward and the architecture we are usually not fond of this. Especially given that you have to break the output to get the correct prediction.  
Might be better to have a new `OwlViTForObjectDetection` no?

File: src/transformers/models/owlvit/convert_owlvit_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        val = val.T
+
+    if new.endswith("bias"):
+        val = val.reshape(-1)
+
+    dct[new] = torch.from_numpy(np.array(val))
+
+
+@torch.no_grad()
+def convert_owlvit_checkpoint(model_name, pytorch_dump_folder_path, push_to_hub):
```
Comment (ArthurZucker) [inline src/transformers/models/owlvit/convert_owlvit_to_hf.py:209]: Same comment about converting the checkpoints, the tests should be optional and outside the conversion in a seperate function if absolutely important.

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+        self.rescale_factor = rescale_factor
+        self.do_pad = do_pad
+        self.do_resize = do_resize
+        self.size = size if size is not None else {"height": 960, "width": 960}
+        self.resample = resample
+        self.do_normalize = do_normalize
+        self.image_mean = image_mean if image_mean is not None else OPENAI_CLIP_MEAN
+        self.image_std = image_std if image_std is not None else OPENAI_CLIP_STD
+
+    def pad(
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:209]: pretty sure we already have this, missing copied from.

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+        **kwargs,
+    ) -> np.ndarray:
+        """
+        Resize an image as per the original implementation.
+
+        Args:
+            image (`np.ndarray`):
+                Image to resize.
+            size (`Dict[str, int]`):
+                ...
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:257]: Todo ?

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+        zoom_factors = [1 / f for f in factors]
+        out = ndi.zoom(filtered, zoom_factors, order=order, mode=ndi_mode, cval=cval, grid_mode=True)
+
+        _clip_warp_output(image, out, mode, cval)
+
+        out = to_channel_dimension_format(out, input_data_format, ChannelDimension.LAST)
+        image = to_channel_dimension_format(out, data_format, input_data_format) if data_format is not None else out
+        return image
+
+    def preprocess(
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:301]: missing some copied from as well no? 
The resize nicely wraps the aliasing with kwargs so the core is pretty similar

File: src/transformers/__init__.py [hunk 9]
```diff
@@ -974,6 +975,7 @@
     _import_structure["models.mobilenet_v2"].extend(["MobileNetV2FeatureExtractor", "MobileNetV2ImageProcessor"])
     _import_structure["models.mobilevit"].extend(["MobileViTFeatureExtractor", "MobileViTImageProcessor"])
     _import_structure["models.oneformer"].extend(["OneFormerImageProcessor"])
+    _import_structure["models.owlv2"].append("Owlv2ImageProcessor")
```
Comment (ArthurZucker) [inline src/transformers/__init__.py:978]: Nicely protected the import

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 10]
```diff
@@ -1325,7 +1328,8 @@ def __init__(self, config: OwlViTConfig):
 
         self.owlvit = OwlViTModel(config)
         self.class_head = OwlViTClassPredictionHead(config)
-        self.box_head = OwlViTBoxPredictionHead(config)
+        self.box_head = OwlViTBoxPredictionHead(config, out_dim=4)
```
Comment (ArthurZucker) [inline src/transformers/models/owlvit/modeling_owlvit.py:1331]: totally ok with this ! Better not to hardcode

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot for the model addition.
I don't think we should change the `OwlViTBoxPredictionHead`, (might be convenient but no it's not transformers philosophy) creating a new class similar to `OwlViTForObjectDetection` like `OwlViT2ForObjectDetection` might be a solution. 
Since that is the only difference, adding a new model and use full copied from does not seem that right to me either

File: src/transformers/models/owlvit/modeling_owlvit.py (hunk 4 above)
Comment (ydshieh) [reply src/transformers/models/owlvit/modeling_owlvit.py:1332]: The change will only happen if `config.add_objectness_head` is `True`. Could you elaborate a bit more about why this is not good?

Comment (ArthurZucker) [reply src/transformers/models/owlvit/modeling_owlvit.py:1332]: It's typical of what we have been trying to avoid when people add models or want specific usages. The previous model architecture will change based on a new config argument. For starters it's the philosophy of not being a toolbox, thus not accepting small changes like this because we could say "why not add an option for a layer norm for people who use layer norm" and etc etc. If we agree for this one, it's an open door for more changes of this kind and becomes hard to not become bloated. 
Second it's not the same `original` model, so less readable for people who only want the OwlVit. In this case, OwlVit2 is from the same family so adding a new OwlVit2ForObjectDetection is what feels the most natural. 
An nice example is for example UMT5: it's basically T5 but we could change the config and the forward to support relative position embeding in a backward compatible way. But we are polluting the original T5 code with unrelated changes. Event if it's the same family, it's a completely different model.

**Comment by NielsRogge** [comment]
Thanks for the reviewing, opened a new PR that adds the model as a new standalone model.

**Comment by BIGBALLON** [comment]
Hi, @NielsRogge，thanks for your great work.
I have a simple question:  
How to convert pytorch model into jax model? Do you have any available tools or scripts?
Thank you very much
